File: solutions/elixir/meetup/1/lib/meetup.ex
defmodule Meetup do
  @moduledoc """
  Calculate meetup dates.
  """

  @type weekday ::
          :monday
          | :tuesday
          | :wednesday
          | :thursday
          | :friday
          | :saturday
          | :sunday

  @type schedule :: :first | :second | :third | :fourth | :last | :teenth

  @weekdays [:monday, :tuesday, :wednesday, :thursday, :friday, :saturday, :sunday]

  @doc """
  Calculate a meetup date.

  The schedule is in which week (1..4, last or "teenth") the meetup date should
  fall.
  """
  @spec meetup(pos_integer, pos_integer, weekday, schedule) :: Date.t()
  def meetup(year, month, weekday, schedule) do
    Date.new(year, month, 1)
    |> then(fn {:ok, date} -> Date.range(date, Date.end_of_month(date)) end)
    |> find_date(order(weekday), schedule)
  end

  defp find_date(days, weekday, :first), do: Enum.find(days, &(Date.day_of_week(&1) == weekday))
  defp find_date(days, weekday, :second), do: find_date(days, weekday, :first) |> Date.shift(week: 1)
  defp find_date(days, weekday, :third), do: find_date(days, weekday, :first) |> Date.shift(week: 2)
  defp find_date(days, weekday, :fourth), do: find_date(days, weekday, :first) |> Date.shift(week: 3)
  defp find_date(days, weekday, :last), do: Enum.reverse(days) |> find_date(weekday, :first)
  defp find_date(days, weekday, :teenth), do: Enum.find(days, fn %Date{day: day} = date -> Date.day_of_week(date) == weekday and day in 13..19 end)

  Enum.each(Enum.with_index(@weekdays, 1), fn {weekday, order} ->
    defp order(unquote(weekday)), do: unquote(order)
  end)
end

File: solutions/erlang/meetup/1/src/meetup.erl
-module(meetup).

-export([meetup/4]).

meetup(Year, Month, DayOfWeek, Week) -> 
    LastDay = last_day_of_month({Year, Month, 1}),
    DateRange = lists:map(fun (Day) -> {Year, Month, Day} end, lists:seq(1, LastDay)),
    find_date(DateRange, order(DayOfWeek), Week).

find_date(Days, WeekDay, first) -> find(Days, fun(Date) -> calendar:day_of_the_week(Date) == WeekDay end);
find_date(Days, WeekDay, second) -> 
    {Year, Month, Day} = find_date(Days, WeekDay, first),
    {Year, Month, Day + 7};
find_date(Days, WeekDay, third) -> 
    {Year, Month, Day} = find_date(Days, WeekDay, first),
    {Year, Month, Day + 14};
find_date(Days, WeekDay, fourth) -> 
    {Year, Month, Day} = find_date(Days, WeekDay, first),
    {Year, Month, Day + 21};
find_date(Days, WeekDay, last) -> find_date(lists:reverse(Days), WeekDay, first);
find_date(Days, WeekDay, teenth) -> find(Days, 
                                        fun({_, _, Day} = Date) -> 
                                            calendar:day_of_the_week(Date) == WeekDay andalso lists:member(Day, lists:seq(13, 19))
                                        end).

order(monday) -> 1;
order(tuesday) -> 2;
order(wednesday) -> 3;
order(thursday) -> 4;
order(friday) -> 5;
order(saturday) -> 6;
order(sunday) -> 7.

find([Item | List], Predicate) ->
    Result = Predicate(Item),
    if
        Result -> Item;
        true -> find(List, Predicate)
    end;
find([], _) -> nil.

last_day_of_month({Year, Month, _}) -> find([31, 30, 29, 28], fun(MaybeLastDay) -> calendar:valid_date({Year, Month, MaybeLastDay}) end).